src/components/TransactionItemRow/TransactionItemRowRBRWithOnyx.tsx

```diff
@@ -0,0 +1,111 @@
+import RenderHTML from '@components/RenderHTML';
+import useLocalize from '@hooks/useLocalize';
+import usePaginatedReportActions from '@hooks/usePaginatedReportActions';
+import useTheme from '@hooks/useTheme';
+import useThemeStyles from '@hooks/useThemeStyles';
+import useTransactionViolations from '@hooks/useTransactionViolations';
+import {isReceiptError} from '@libs/ErrorUtils';
+import {getIOUActionForTransactionID} from '@libs/ReportActionsUtils';
+import ViolationsUtils from '@libs/Violations/ViolationsUtils';
+import variables from '@styles/variables';
+import type {Errors} from '@src/types/onyx/OnyxCommon';
+import type ReportAction from '@src/types/onyx/ReportAction';
+import type Transaction from '@src/types/onyx/Transaction';
+import type {ReceiptError, ReceiptErrors} from '@src/types/onyx/Transaction';
+
+type TransactionItemRowRBRProps = {
+    /** Transaction item */
+    transaction: Transaction;
+
+    /** Styles for the RBR messages container */
+    containerStyles?: ViewStyle[];
+};
+
+/**
+ * Extracts unique error messages from errors and actions
+ */
+const extractErrorMessages = (errors: Errors | ReceiptErrors, errorActions: ReportAction[], translate: LocaleContextProps['translate']): string[] => {
+    const uniqueMessages = new Set<string>();
+
+    // Combine transaction and action errors
+    let allErrors: Record<string, string | Errors | ReceiptError | null | undefined> = {...errors};
+    errorActions.forEach((action) => {
+        if (!action.errors) {
+            return;
+        }
+        allErrors = {...allErrors, ...action.errors};
+    });
+
+    // Extract error messages
+    Object.values(allErrors).forEach((errorValue) => {
+        if (!errorValue) {
+            return;
+        }
+        if (typeof errorValue === 'string') {
+            uniqueMessages.add(errorValue);
+        } else if (isReceiptError(errorValue)) {
+            uniqueMessages.add(translate('iou.error.receiptFailureMessageShort'));
+        } else {
+            Object.values(errorValue).forEach((nestedErrorValue) => {
+                if (!nestedErrorValue) {
+                    return;
+                }
+                uniqueMessages.add(nestedErrorValue);
+            });
+        }
+    });
+
+    return Array.from(uniqueMessages);
+};
+
+function TransactionItemRowRBRWithOnyx({transaction, containerStyles}: TransactionItemRowRBRProps) {
+    const styles = useThemeStyles();
+    const transactionViolations = useTransactionViolations(transaction?.transactionID);
+    const {translate} = useLocalize();
+    const theme = useTheme();
+
+    const {sortedAllReportActions: transactionActions} = usePaginatedReportActions(transaction.reportID);
+    const transactionThreadId = transactionActions ? getIOUActionForTransactionID(transactionActions, transaction.transactionID)?.childReportID : undefined;
+    const {sortedAllReportActions: transactionThreadActions} = usePaginatedReportActions(transactionThreadId);
```

Comment:
Could we move this logic up in the tree to the main list and do all those computations for all rows there once and just pass the messages for each row (so we dont have to repeat same computations in each row)